### PR TITLE
use ui-components header in GRC

### PR DESCRIPTION
**Pull request by willkutler**
- modifies GRC to use AcmHeader from ui-components
- adds routes needed by the header for the welcome page, ACM version, username, logout, etc

for https://github.com/open-cluster-management/backlog/issues/8781

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -35,7 +35,7 @@ module.exports = {
     warningsFilter: [/Failed to parse source map/],
   },
   entry: {
-    'main': ['@babel/polyfill', './src-web/index.js']
+    'main': ['@babel/polyfill', './src-web/index.js', './scss/common.scss', './scss/welcome.scss'],
```
Comment (ycao56) [inline webpack.config.js:38]: I wonder if this is still needed with the welcome page change

File: views/layouts/main.handlebars [hunk 2]
```diff
@@ -5,9 +5,9 @@
 <head>
     <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
     <title>{{title}}</title>
-    <link rel='stylesheet' href="{{filesH.css.path}}">
     <link rel='stylesheet' href="{{contextPath}}/{{manifest.main.css}}">
-    <link rel='icon' href="{{headerContextPath}}/graphics/RHACM-Favicon.svg">
+    <link rel='stylesheet' href="{{contextPath}}/{{manifest.welcome.css}}">
```
Comment (ycao56) [inline views/layouts/main.handlebars:9]: I wonder if this is still needed with the welcome page change

File: views/home.handlebars [hunk 3]
```diff
@@ -1,18 +1,18 @@
...
     <div id="page">{{{content}}}</div>
     {{{body}}}
-    <script charset="UTF-8">window.__PRELOADED_STATE__= {{{json stateH}}}</script>
+    {{!-- <script charset="UTF-8">window.__PRELOADED_STATE__= {{{json stateH}}}</script> --}}
     <script id='config' type='application/json'>{{{json config}}}</script>
     <script id='context' type='application/json'>{{{json context}}}</script>
     <script id='propshcm' type='application/json'>{{{json props}}}</script>
     <script src="{{contextPath}}/{{manifest.vendorhcm.js}}"></script>
-    <script id='props' type='application/json'>{{{json propsH}}}</script>
+    {{!-- <script id='props' type='application/json'>{{{json propsH}}}</script>
```
Comment (ycao56) [inline views/home.handlebars:12]: clean up

File: views/home.handlebars [hunk 4]
```diff
@@ -1,18 +1,18 @@
 {{!-- Copyright (c) 2020 Red Hat, Inc. --}}
 
 <body class="bx--global-light-ui" aria-label={{title}}>
-    <div id="header">{{{header}}}</div>
+    {{!-- <div id="header">{{{header}}}</div> --}}
```
Comment (ycao56) [inline views/home.handlebars:4]: clean up

File: views/home.handlebars [hunk 5]
```diff
@@ -1,18 +1,18 @@
...
     <div id="page">{{{content}}}</div>
     {{{body}}}
-    <script charset="UTF-8">window.__PRELOADED_STATE__= {{{json stateH}}}</script>
+    {{!-- <script charset="UTF-8">window.__PRELOADED_STATE__= {{{json stateH}}}</script> --}}
```
Comment (ycao56) [inline views/home.handlebars:7]: clean up

File: app.js [hunk 6]
```diff
@@ -92,7 +92,7 @@ function setReq(req) {
 }
 
 const { createProxyMiddleware } = require('http-proxy-middleware')
-app.use(`${appConfig.get('contextPath')}/graphql`, cookieParser(), csrfMiddleware, (req, res, next) => {
+app.use(`${appConfig.get('contextPath')}/graphql`, cookieParser(), (req, res, next) => {
```
Comment (ycao56) [inline app.js:95]: add csrfMiddleware back. https://github.com/open-cluster-management/grc-ui/pull/431/files#diff-dbed998603a94ea423636bb9aa34e71e6253bb771a3d7b51311f17a741294c1eR5 should add correct token to the ui

File: views/home.handlebars [hunk 7]
```diff
@@ -1,18 +1,19 @@
 {{!-- Copyright (c) 2020 Red Hat, Inc. --}}
 
 <body class="bx--global-light-ui" aria-label={{title}}>
-    <div id="header">{{{header}}}</div>
+    {{!-- <div id="header">{{{header}}}</div> --}}
+    <input type="hidden" id="app-access" value={{props.xsrfToken}}>
```
Comment (ycao56) [inline views/home.handlebars:5]: also need to remove from src-web/containers/App.js

File: controllers/ui/app.js [hunk 8]
```diff
@@ -11,118 +11,202 @@
...
+    })
   })
+})
+
+function fetchHeader(req, res, store, ctx) {
+
+  res.render('home', Object.assign({
+    manifest: appUtil.app().locals.manifest,
+    contextPath: config.get('contextPath'),
+    headerContextPath: config.get('headerContextPath'),
```
Comment (ycao56) [inline controllers/ui/app.js:198]: headerContextPath variable should no longer needed. clean it up in our code

File: controllers/ui/app.js [hunk 9]
```diff
@@ -11,118 +11,202 @@
...
+
+      App = App === undefined ? require('../../src-web/containers/App').default : App
+
+      const fetchHeaderContext = getContext(req)
+      return fetchHeader(req, res, store, fetchHeaderContext)
+    })
   })
+})
+
+function fetchHeader(req, res, store, ctx) {
```
Comment (ycao56) [inline controllers/ui/app.js:193]: this entire function seems not necessary. can be merged to where it is called

File: lib/client/apollo-client.js [hunk 10]
```diff
@@ -130,6 +130,11 @@ class GrcApolloClient {
     return Promise.resolve()
   }
 
+  //gets user access data
+  getUserAccess(variables = {}) {
+    return client.query({ query: Query.getUserAccess, variables: { variables }})
```
Comment (ycao56) [inline lib/client/apollo-client.js:135]: redundant `variables`

File: webpack.dll.js [hunk 11]
```diff
@@ -52,7 +52,7 @@ module.exports = {
           'css-loader',],
       },
       {
-        test: /\.(woff2?|ttf|eot|otf)(\?.*$|$)/,
+        test: /\.(png|jpg|jpeg|gif|svg|woff2?|ttf|eot|otf)(\?.*$|$)/,
```
Comment (ycao56) [inline webpack.dll.js:55]: is this change necessary?

File: lib/server/userInfo.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+      serviceaccountToken = fs.readFileSync(`${serviceAccountPath}/token`, 'utf8')
+    } else {
+      serviceaccountToken = process.env.SERVICEACCT_TOKEN || ''
+    }
+  } catch (err) {
+    logger.error('Error reading service account token', err && err.message)
+  }
+
+  const options = {
+    url: `${config.get('API_SERVER_URL')}/apis/authentication.k8s.io/v1/tokenreviews`,
```
Comment (ycao56) [inline lib/server/userInfo.js:26]: Should be able to move this call to /console, can do it in a separate PR

File: lib/server/oauth-info-client.js [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+      serviceaccountToken = process.env.SERVICEACCT_TOKEN || ''
+    }
+  } catch (err) {
+    logger.error('Error reading service account token', err && err.message)
+  }
+
+  // for both running local and inside cluster, the below endpoint can be used to
+  // get the oauth2 server and authorizepath  and tokenpath endpoints.
+  const options = {
+    url: `${config.get('API_SERVER_URL')}/.well-known/oauth-authorization-server`,
```
Comment (ycao56) [inline lib/server/oauth-info-client.js:24]: Should be able to move this call to /console, can do it in a separate PR

File: src-web/containers/App.js [hunk 14]
```diff
@@ -51,6 +57,12 @@ class App extends React.Component {
     return this.props.staticContext
   }
 
+  componentDidMount() {
+    GrcApolloClient.getUserAccess().then((response) => {
```
Comment (ycao56) [inline src-web/containers/App.js:61]: should handle error

Comment (ycao56) [reply src-web/containers/App.js:61]: This call should be in `src-web/actions/common.js`

File: lib/client/apollo-client.js [hunk 15]
```diff
@@ -130,6 +130,11 @@ class GrcApolloClient {
     return Promise.resolve()
   }
 
+  //gets user access data
+  getUserAccess(variables = {}) {
```
Comment (ycao56) [inline lib/client/apollo-client.js:134]: `variables = {}` this is also redundant

File: lib/client/apollo-client.js [hunk 16]
```diff
@@ -130,6 +130,11 @@ class GrcApolloClient {
...
+    return client.query({ query: Query.getUserAccess, variables })
```
Comment (ycao56) [inline lib/client/apollo-client.js:135]: same here

File: src-web/actions/access.js [hunk 17]
```diff
@@ -4,10 +4,24 @@
...
+import GrcApolloClient from '../../lib/client/apollo-client'
 
 export const userAccessSuccess = access => ({
   type: USER_ACCESS_SUCCESS,
   access
 })
 
 export const userAccessFailure = err => ({ type: USER_ACCESS_FAILURE, err })
+
+export const getUserAccessData = () => {
```
Comment (ycao56) [inline src-web/actions/access.js:16]: should be in src-web/actions/common.js

**Review by ycao56 — APPROVED** [review]

**Comment by willkutler** [comment]
https://github.com/open-cluster-management/grc-ui/pull/431 resolves this
